File: code-challenges/pure-code/01-lru-cache/solution.optimized.py
"""
LRU Cache — solución optimizada (Python)

OrderedDict preserva orden de inserción:
  - Primer key = LRU, último = MRU
  - get/put existentes: move_to_end(key) → MRU O(1)
  - Evict: popitem(last=False) → saca el LRU O(1)

Complejidad: get/put O(1) amortizado | espacio O(capacity)
Alternativa clásica: dict + doubly linked list (misma complejidad).
Thread-safety: sin lock, get/put concurrentes corrompen orden/size.
"""

from __future__ import annotations

import json
import sys
from collections import OrderedDict
from pathlib import Path


class LRUCache:
    def __init__(self, capacity: int) -> None:
        if capacity < 1:
            raise ValueError("capacity must be >= 1")
        self.capacity = capacity
        self._map: OrderedDict[int, int] = OrderedDict()

    def get(self, key: int) -> int:
        if key not in self._map:
            return -1
        self._map.move_to_end(key)  # MRU
        return self._map[key]

    def put(self, key: int, value: int) -> None:
        if key in self._map:
            self._map.move_to_end(key)
            self._map[key] = value
            return
        if len(self._map) >= self.capacity:
            self._map.popitem(last=False)  # evict LRU
        self._map[key] = value


def run_case(case: dict) -> dict:
    cache: LRUCache | None = None
    got: list[int] = []

    for op in case["ops"]:
        name = op[0]
        if name == "LRUCache":
            cache = LRUCache(op[1])
        elif name == "put":
            cache.put(op[1], op[2])  # type: ignore[union-attr]
        elif name == "get":
            got.append(cache.get(op[1]))  # type: ignore[union-attr]
        else:
            raise ValueError(f"Unknown op: {name}")

    expected = case["expected"]
    passed = got == expected
    return {"name": case["name"], "passed": passed, "got": got, "expected": expected}


def main() -> None:
    # 1. Determina la ruta del archivo de casos de prueba.
    #   - Si el usuario pasó un argumento al script (sys.argv[1]), lo usa como ruta.
    #   - Si no, usa "starter/cases.json" como archivo de entrada por defecto.
    path = Path(sys.argv[1] if len(sys.argv) > 1 else "starter/cases.json")

    # 2. Lee el contenido del archivo y lo parsea como JSON.
    #   - path.read_text(...): Lee el archivo como texto (asumiendo encoding UTF-8).
    #   - json.loads(...): Convierte el texto leído en un objeto Python (dict).
    raw = json.loads(path.read_text(encoding="utf-8"))

    # 3. Resuelve todos los casos de prueba.
    #   - Para cada caso en raw["cases"], llama a run_case(c) para obtener el resultado de ese caso.
    #   - results será una lista de diccionarios con los detalles de cada caso (nombre, si pasó, esperado, obtenido).
    results = [run_case(c) for c in raw["cases"]]

    # 4. Verifica si TODOS los casos de prueba pasaron.
    #   - Si todos los resultados tienen r["passed"] igual a True, passed será True.
    passed = all(r["passed"] for r in results)

    # 5. Imprime un resumen de los resultados en JSON, formateado con indentación.
    #   - Incluye si todos pasaron y los detalles de cada caso.
    print(json.dumps({"passed": passed, "results": results}, indent=2))

    # 6. Sale del programa con código 0 si todos los tests pasaron, o 1 si alguno falló.
    #   - Esto ayuda a integrarlo con sistemas de CI o scripts de testing.
    sys.exit(0 if passed else 1)

if __name__ == "__main__":
    main()

File: code-challenges/pure-code/01-lru-cache/solution.optimized.ts
/**
 * LRU Cache — versión optimizada
 *
 * ---------------------------------------------------------------------------
 * QUÉ FALLABA EN solution.ts (tu versión)
 * ---------------------------------------------------------------------------
 * 1) CORRECTITUD — put no actualiza claves existentes
 *    Siempre hace push(). En "update_existing": put(1,9) con capacity 2
 *    debería reemplazar el valor de 1 y refrescar recencia, NO añadir un
 *    tercer elemento ni desalojar a ciegas.
 *
 * 2) CORRECTITUD — put a capacidad llena siempre llama freeOldest()
 *    Aunque la key ya exista. Actualizar no debe evictar.
 *
 * 3) RENDIMIENTO — get/put hacen findIndex O(n) + sort O(n log n)
 *    El objetivo del challenge es O(1) amortizado por operación.
 *    Con 50k ops, sort en cada put/get se nota.
 *
 * 4) TIPADO — value: any (rompe la convención strict del challenge)
 *
 * 5) Detalle menor — get miss: return index (-1) funciona de casualidad;
 *    es más claro return -1 explícito.
 *
 * ---------------------------------------------------------------------------
 * POR QUÉ ESTA VERSIÓN ES MEJOR
 * ---------------------------------------------------------------------------
 * Map en JS/TS preserva orden de inserción:
 *   - Primer key = least recently used (LRU)
 *   - Último key = most recently used (MRU)
 * Para “tocar” una key: delete + set → queda al final (MRU) en O(1).
 * Evict: borrar map.keys().next().value (el primero) en O(1).
 *
 * Complejidad: get/put → O(1) amortizado | espacio → O(capacity)
 *
 * Alternativa “de entrevista clásica”: HashMap + Doubly Linked List
 * (misma complejidad; más código). Map ordenado es el idiomático en TS.
 * ---------------------------------------------------------------------------
 */

import fs from "node:fs";

class LRUCache {
  private readonly capacity: number;
  private readonly map = new Map<number, number>();

  constructor(capacity: number) {
    if (capacity < 1) {
      throw new Error("capacity must be >= 1");
    }
    this.capacity = capacity;
  }

  get(key: number): number {
    if (!this.map.has(key)) {
      return -1;
    }
    const value = this.map.get(key)!;
    // Refresh recency: move to MRU (end of Map insertion order)
    this.map.delete(key);
    this.map.set(key, value);
    return value;
  }

  put(key: number, value: number): void {
    if (this.map.has(key)) {
      // Update existing: remove old position, re-insert as MRU
      this.map.delete(key);
    } else if (this.map.size >= this.capacity) {
      // Evict LRU = first key in insertion order
      const oldestKey = this.map.keys().next().value as number;
      this.map.delete(oldestKey);
    }
    this.map.set(key, value);
  }
}

// --- Runner (valida contra cases.json) -------------------------------------

type Op = [string, ...number[]];

interface Case {
  name: string;
  ops: Op[];
  expected: number[];
}

function runCase(c: Case): { name: string; passed: boolean; got: number[]; expected: number[] } {
  let cache: LRUCache | undefined;
  const got: number[] = [];

  for (const op of c.ops) {
    const [name, a, b] = op;
    switch (name) {
      case "LRUCache":
        cache = new LRUCache(a);
        break;
      case "put":
        cache!.put(a, b);
        break;
      case "get":
        got.push(cache!.get(a));
        break;
      default:
        throw new Error(`Unknown op: ${name}`);
    }
  }

  const passed =
    got.length === c.expected.length && got.every((v, i) => v === c.expected[i]);

  return { name: c.name, passed, got, expected: c.expected };
}

function main(): void {
  const path = process.argv[2] ?? "starter/cases.json";
  const raw = JSON.parse(fs.readFileSync(path, "utf8")) as { cases: Case[] };
  const results = raw.cases.map(runCase);
  const passed = results.every((r) => r.passed);

  console.log(JSON.stringify({ passed, results }, null, 2));
  process.exit(passed ? 0 : 1);
}

main();
